pandas/_libs/lib.pyx

```diff
@@ -1353,6 +1383,67 @@ cdef class Seen:
             self.datetime_ or self.datetimetz_ or self.nat_ or self.timedelta_
             or self.period_ or self.interval_ or self.numeric_ or self.object_
         )
+    cdef int get_curr_val(self, curr_seen_t *ret_val) except -1:
```

Comment:
This function acts as a setter not a getter. Probably better named as `set_current_value`

Reply:
Well, technically it gets the value from the Seen object. Assigning to the pointer is awkward, but it's better than having a dummy parameter just for the sake of dispatching correctly.

Maybe there's a better style for doing this though?

Reply:
I think you can still make this a getter if you remove the pointer reference and just do `complexes[i] = self.get_curr_val()`

Reply:
Sorry, I don't think I was being clear/thinking clearly when I typed that. I'll write a more in-depth explanation below.

To recap, the current approach requires a pointer since I am currently using a fused type(to reduce code duplication).
In Cython, fused functions need to have a fused type as a parameter, so that Cython knows which variant to dispatch to. So, our function signature has to look something like this
```
cdef int get_curr_val(self, curr_seen_t *ret_val) except -1: # Current
cdef curr_seen_t get_curr_val(self, curr_seen_t *ret_val): # If we were to return curr_seen_t, we lose error handling here
```

While thinking about what you said, though, maybe there is a third approach using void pointers.
Then, our signature would look like this
```
cdef void* get_curr_val(self, CurrSeen ret_type)
```
Here, by using void* instead of a fused function, we avoid having to pass a curr_seen_t pointer/regular type parameter into the function.

This function would then get the currently seen value, and then cast it to the return type (while checking that we are doing a correct cast like before), and return a pointer to it in the form of a void pointer.

I think we would be able to keep the error checking here (by using NULL as the error value).
Then, we would do e.g. 
```
complexes[i] = ((complex *)seen.get_curr_val())[0]
```
This is a lot more uglier than ``seen.get_curr_val(&complexes[i])``, but maybe it is better C style (it also gets rid of the fused functions) ?

Reply:
I'm not sure I follow why you would lose error handling when returning `curr_seen_t`. If you use `except *` it should add a `PyErr_Occurred()` check after each invocation. That adds a performance penalty compared to a more targeted approach, but not sure how much it would be in the grand scheme of this refactor

Reply:
Sorry, I forgot about that case (would still like to avoid, though).

I think my point about the dummy parameters still stands, though.

Reply:
I think it would be simpler if you just created `get_{int|uint|float|complex}_value` functions. I know it seems  basic compared to fused types but injecting an arbitrary parameter to appease the templating requirements really hurts the readability.

Reply:
I've updated the code to not use the fused function now. Performance seems to be unaffected.